c++/test/TestReader.cc

```diff
@@ -250,4 +252,206 @@ namespace orc {
+    // select all of single_map and only the outermost offsets of nested_map.
+    verifyMapSelection({{2, ReadIntent_ALL}, {5, ReadIntent_OFFSETS}}, {0, 2, 3, 4, 5});
+  }
+
+  TEST(TestReadIntent, testMapConflictingIntent) {
+    // test conflicting ReadIntent on nested_map.
+    verifyMapSelection({{5, ReadIntent_OFFSETS}, {9, ReadIntent_ALL}}, {0, 5, 7, 9, 10, 11});
+    verifyMapSelection({{5, ReadIntent_ALL}, {9, ReadIntent_OFFSETS}}, {0, 5, 6, 7, 8, 9, 10, 11});
+    verifyMapSelection({{5, ReadIntent_OFFSETS}, {7, ReadIntent_ALL}, {9, ReadIntent_OFFSETS}},
+                       {0, 5, 7, 8, 9, 10, 11});
+  }
+
+  TEST(TestReadIntent, testMapRowBatchContent) {
+    std::unique_ptr<Reader> reader = createExampleReader("TestOrcFile.nestedMap.orc");
+
+    // select all of single_map and only the offsets of nested_map.
+    RowReaderOptions::IdReadIntentMap idReadIntentMap = {{2, ReadIntent_ALL},
+                                                         {5, ReadIntent_OFFSETS}};
+    RowReaderOptions rowReaderOpts;
+    rowReaderOpts.includeTypesWithIntents(idReadIntentMap);
+    std::unique_ptr<RowReader> rowReader = reader->createRowReader(rowReaderOpts);
+
+    // Read a row batch.
+    std::unique_ptr<ColumnVectorBatch> batch = rowReader->createRowBatch(1024);
+    EXPECT_TRUE(rowReader->next(*batch));
+    EXPECT_EQ(1, batch->numElements);
+    auto& structBatch = dynamic_cast<StructVectorBatch&>(*batch);
+
+    // verify content of int_array selection.
+    auto& mapBatch = dynamic_cast<MapVectorBatch&>(*structBatch.fields[0]);
+    auto& keyBatch = dynamic_cast<StringVectorBatch&>(*mapBatch.keys);
+    auto& valueBatch = dynamic_cast<StringVectorBatch&>(*mapBatch.elements);
+    EXPECT_EQ(1, mapBatch.numElements);
+    EXPECT_EQ(0, mapBatch.offsets.data()[0]);
+    EXPECT_EQ(1, mapBatch.offsets.data()[1]);
+    // verify key content.
+    EXPECT_EQ(1, keyBatch.numElements);
+    EXPECT_EQ(2, keyBatch.length.data()[0]);
+    EXPECT_EQ(0, strncmp("k0", keyBatch.data.data()[0], 2));
+    // verify value content.
+    EXPECT_EQ(1, valueBatch.numElements);
+    EXPECT_EQ(2, valueBatch.length.data()[0]);
+    EXPECT_EQ(0, strncmp("v0", valueBatch.data.data()[0], 2));
+
+    // verify content of nested_map selection.
+    auto& nestedMapBatch = dynamic_cast<MapVectorBatch&>(*structBatch.fields[1]);
+    EXPECT_EQ(1, nestedMapBatch.numElements);
+    EXPECT_EQ(0, nestedMapBatch.offsets.data()[0]);
+    EXPECT_EQ(1, nestedMapBatch.offsets.data()[1]);
+    EXPECT_EQ(nullptr, nestedMapBatch.keys.get());
+    EXPECT_EQ(nullptr, nestedMapBatch.elements.get());
+  }
+
+  /**
+   * Read TestOrcFile.nestedUnion.orc and verify the resolved selections.
+   *
+   * The ORC fie has the following schema:
+   * struct<
+   *   id:int,
+   *   single_union:uniontype<int,string>,
+   *   nested_union:uniontype<uniontype<int,uniontype<int,string>>,int>
+   * >
+   */
+  void verifyUnionSelection(const RowReaderOptions::IdReadIntentMap& idReadIntentMap,
```

Comment:
The same as `verifyMapSelection`, can we merge this function with `verifySelection`?

Reply:
Done